extension.md

```diff
@@ -0,0 +1,28 @@
+## Further Analysis
+
+### Economics of Cross-Witness
+
+If any participant cheats, his money at stake must be enough to cover for all the expenses. The contract denomination must be grater than aggregated transaction costs of all the participants, plus potential arbitration costs. The aggregated costs are proportionate to the number of participants, and therefore there must be a limit to the number of participants for each denomination of smart contract. If the number of people exceeds the contract limit, it might result in shortages/queues. This gives the malicious entities an opportunity to launch DDoS attack. One of the possible solutions: multiple contracts + one "router" contract, through which all the TX must be initiated.
+
+### Random Numbers Generation (RNG)
+
+How can we prove, that router contract is honest ("nothing up my sleeve"), and besides, may not be manipulated by miners or anyone else?
+After the contract is activated, everybody must sign the next block's hash with his public key, and then publish the signature. Those signatures, if valid, will be in turn concatenated or XORed, and the result hashed, thus producing the seed for RNG.
+
+BTW, this algorithm might be also used in gambling dapps. They might even call it "rolling the dice" for marketing purposes, as it gives the participants the same illusion as physically rolling the dice -- illusion of some kind of personal control over the RNG.
+
+### Onion Concept
+
+We assume, that multiple iterations will be necessary to acheive the desired probability of untracebility. If we could do several iterations at once, we might save on transaction costs, execution time, and -- which might be even more important -- mental costs of initiating transaction, choosing the available contract in case of queues, and managing the multiple accounts (although this could be automated, but still requires external overhead).
+
+Here is one of the possible solutions. During the Witness Selection step, everybody should select certain (predefined) number of witnesses, then repeatedly encrypt his data with their public keys, then publish the resulting encrypted data ("onion"). Next step remains the same, except that it must be repeated as many times, as required to "unwind the onion".
```

Comment:
yeah - i considered this myself. the multiple rounds may end up being too onerous for most uses though; rounds are expensive as they require all participants to SSTORE and would likely take a safe minimum of 2 minutes (transaction propagation and block inclusion often takes > 4 blocks).